### PR TITLE
Rewrite models based on Square struct, add LineMPO struct

### PR DESCRIPTION
**Pull request by LinjianMa**
Change the hamiltonians in the PEPS optimization experiments such that each local hamiltonian is a column/row of the overall H.

### CODE REVIEW

File: test/Optimizations/runtests.jl [hunk 1]
```diff
@@ -29,6 +29,19 @@ using ITensorNetworkAD.ITensorAutoHOOT: batch_tensor_contraction
   end
 end
 
+@testset "test the equivalence of local and line hamiltonian" begin
```
Comment (mtfishman) [inline test/Optimizations/runtests.jl:32]: We could also add a test here comparing the optimized energy against DMRG or exact diagonalization (maybe there is another test of that).

File: src/ITensorNetworks/lattices.jl [hunk 2]
```diff
@@ -90,6 +106,20 @@ function incident_edges(lattice::HyperCubic{N}, site::NTuple{N,Int}) where {N}
...
 
-function bonds(lattice::HyperCubic)
-  return ((s, n) for s in sites(lattice) for n in outneighbors(lattice, s))
+function bonds(lattice::HyperCubic; periodic=false)
+  return [
+    (s, n) for s in sites(lattice) for n in outneighbors(lattice, s; periodic=periodic)
+  ]
+end
+
+function bonds(lattice::Square, coord::Tuple{Colon,<:Integer})
```
Comment (mtfishman) [inline src/ITensorNetworks/lattices.jl:115]: This is useful functionality!

In general what we could do is make this more generic, where you could ask for a general (hypercubic) slice of a `HyperCubic` lattice type and then ask for the bonds of that slice. But totally fine to have more specialized functions like this for now.

**Comment by mtfishman** [comment]
Looks good, glad you were able to fix the bond iteration for the `HyperCubic` lattice type.

I think ideally we wouldn't have two different types `LocalMPO` and `LineMPO`, in principle we can merge them into one type where it stores the list or range of sites where that Hamiltonian term is acting. But the current design is fine for now.

We should make issues reminding us to:

1. Merge `LocalMPO` and `LineMPO` into one type.
2. Generalize the version of `bonds` that returns an iterator over the bonds of a hypercubic sublattice of a hypercubic lattice (and consider writing a general `SubHyperCubic` type which is a slice of a `HyperCubic` type, or allow adding slicing information into the `HyperCubic` type itself).

File: test/Optimizations/runtests.jl (hunk 1 above)
Comment (LinjianMa) [reply test/Optimizations/runtests.jl:32]: Are you talking about comparing the energy of gradient based methods with DMRG, or just DMRG with different hamiltonians? I guess for the first one the gradient based method performs worse, so not sure whether that could be added as a test

Comment (mtfishman) [reply test/Optimizations/runtests.jl:32]: I just mean test out that the final energy is correct (how you get the energy is not so important, I just brought up DMRG as an easy method to use to obtain the energy in a different way).

Comment (mtfishman) [reply test/Optimizations/runtests.jl:32]: It could be a very loose test bound (just check that the first 2 digits of the energy are correct).

Comment (LinjianMa) [reply test/Optimizations/runtests.jl:32]: oh OK, yeah that makes sense

Comment (LinjianMa) [reply test/Optimizations/runtests.jl:42]: done

**Pull request merged**